pallets/randomness/src/vrf.rs

```diff
@@ -0,0 +1,126 @@
+// Copyright 2019-2022 PureStake Inc.
+// This file is part of Moonbeam.
+
+// Moonbeam is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+//! VRF logic
+use crate::{Config, CurrentVrfInput, GetVrfInput, LocalVrfOutput, RandomnessResults, RequestType};
+use frame_support::{pallet_prelude::Weight, traits::Get};
+use nimbus_primitives::{NimbusId, NIMBUS_ENGINE_ID};
+use parity_scale_codec::{Decode, Encode};
+use scale_info::TypeInfo;
+#[cfg(feature = "std")]
+use serde::{Deserialize, Serialize};
+pub use session_keys_primitives::make_transcript;
+use session_keys_primitives::{KeysLookup, PreDigest, VrfId, VRF_ENGINE_ID, VRF_INOUT_CONTEXT};
+use sp_consensus_vrf::schnorrkel;
+use sp_core::crypto::ByteArray;
+use sp_runtime::RuntimeDebug;
+
+/// VRF output
+type Randomness = schnorrkel::Randomness;
+
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Default, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+/// VRF inputs from the relay chain
+/// Both inputs are expected to change every block
+pub struct VrfInput<SlotNumber, RelayHash> {
+	/// Relay block slot number
+	pub slot_number: SlotNumber,
+	/// Relay block storage root
+	pub storage_root: RelayHash,
+}
+
+/// Set vrf input in storage and log warning if either of the values did NOT change
+/// Called in previous block's `on_finalize`
+pub(crate) fn set_input<T: Config>() {
+	let input = T::VrfInputGetter::get_vrf_input();
+	if let Some(last_vrf_input) = <CurrentVrfInput<T>>::take() {
+		// logs if input uniqueness assumptions are violated (no reuse of vrf inputs)
+		if last_vrf_input.storage_root == input.storage_root
+			|| last_vrf_input.slot_number == input.slot_number
+		{
+			log::warn!(
```

Comment:
Maybe a code comment about what you're worried might happen (e.g. some other pallet/runtime/whatever failed to do XYZ)